Implements support for ECDSA keys. Fixes #2163.

atombrella opened the pull request:
Continued from #8254

I'm not sure about the conflict with run on existing certificates. I'm willing to have a look at it, but a few hints are appreciated.

I added some code as NamespaceConfig.key_type, but I had to wrap an try-except around it, since the tests were failing without it. Again, please hint if there's a better way to deal with this.

I'm hoping this will eventually be in a release of certbot.

Thanks to @pahrohfit and @Tomoyuki-GH for previous efforts to implement
suport for this.

Co-Authored-By: Robert Dailey <rob@wargam.es>
Co-Authored-By: Tomoyuki-GH <55397638+Tomoyuki-GH@users.noreply.github.com>

## Pull Request Checklist

- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Include your name in `AUTHORS.md` if you like.

adferrand commented:
I commit myself to keep the target branch `ecdsa` in sync with master until this PR is merged.

atombrella commented:
> I commit myself to keep the target branch `ecdsa` in sync with master until this PR is merged.

Thanks. I think I addressed your comments from the original PR :) If not, we can do another round. I think there are also 1-2 issues from the ecdsa area that need to be handled before this lands in master.

adferrand commented on certbot/certbot/_internal/configuration.py:
```diff
@@ -138,6 +142,27 @@ def renewal_post_hooks_dir(self):
                             constants.RENEWAL_POST_HOOKS_DIR)
 
+    @property
+    def key_type(self):
```
I think you do not need this function to compute the `key_type` dynamically in fact. I believe you added it to answer to the comment I wrote in https://github.com/certbot/certbot/pull/8254#discussion_r510807545.

The situation to cover here is a bad behavior between `--force-renewal` and `--reuse-key` when `certbot run` or `certbot certonly` are run. In these precise cases, the renewal configuration parameters related to the key is lost because of the way it is implemented in the `renewal` module.

The following code has been written (and updated by you) to cover this corner case.
https://github.com/atombrella/certbot/blob/aa1950347d225e3067771a40cc397f4af0bcb765/certbot/certbot/_internal/renewal.py#L504-L513

To completely fix the issue (by storing `key_type`, you can simply modify the function `_update_renewal_params_from_key`. What you should to is to set `config.key_type = 'rsa'` in case of an RSA key, and `config.key_type = 'ecdsa'` in case of an ECDSA key.

atombrella replied:
OK. Done.

adferrand reviewed (changes requested):
I think that from my last review (https://github.com/certbot/certbot/pull/8254#pullrequestreview-515564860) the only remaining point is below.

Beside this, what is the problem of conflict with run on existing certificates you are referring to?

atombrella commented:
> Beside this, what is the problem of conflict with run on existing certificates you are referring to?

Maybe I misunderstood the extend of this issue.
https://github.com/certbot/certbot/issues/8365

adferrand commented:
> 
> 
> > Beside this, what is the problem of conflict with run on existing certificates you are referring to?
> 
> Maybe I misunderstood the extend of this issue.
> #8365

Ok this was this issue. No problem, we can tackle that after this PR.

adferrand reviewed (approved):
And this LGTM!

@atombrella, thanks you a lot for your patience and your dedication to make ECDSA certificates live in Certbot.

(I merge this PR instead of squashing to ensure authorship is preserved. It will need to be the same for `ecdsa` to `master`).

HOSTED-POWER commented:
In which certbot version can we expect ECDSA support? :)

adferrand commented:
I have big hopes for the next one.

adferrand commented:
The feature has been merged to master. It will be available for the next Certbot release!
